Add warning for deprecated to_edge() flow with CoreML

jagadeeshkaruturi11 opened the pull request:
This commit adds a warning message in the CoreML partitioner's partition() method to alert users when they use the older to_edge() flow instead of the recommended to_edge_transform_and_lower() flow.

The warning informs users that:
- Using the old flow may result in performance regression
- The recommended approach is to use to_edge_transform_and_lower() with CoreML partitioner
- Provides a link to the CoreML backend documentation for more details

Fixes #15960

cc @kimishpatel @YifanShenSZ @cymbalrush @metascroy

jagadeeshkaruturi11 committed:
Add warning for deprecated to_edge() flow with CoreML

This commit adds a warning message in the CoreML partitioner's partition() method to alert users when they use the older to_edge() flow instead of the recommended to_edge_transform_and_lower() flow.

The warning informs users that:
- Using the old flow may result in performance regression
- The recommended approach is to use to_edge_transform_and_lower() with CoreML partitioner
- Provides a link to the CoreML backend documentation for more details

Fixes #15960

Lokesh9106 commented:
Thank you for your appreciation
Can I make changes and resubmit it

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________
From: Scott Roy ***@***.***>
Sent: Tuesday, November 25, 2025 12:59:05 AM
To: pytorch/executorch ***@***.***>
Cc: BELLAM LOKESH ***@***.***>; Mention ***@***.***>
Subject: Re: [pytorch/executorch] Add warning for deprecated to_edge() flow with CoreML (PR #15963)

[https://avatars.githubusercontent.com/u/161522778?s=20&v=4]metascroy left a comment (pytorch/executorch#15963)<https://github.com/pytorch/executorch/pull/15963#issuecomment-3572388215>

Great start! But I'm afraid adding a warning inside the partitioner itself isn't sufficient because the partitioner is also used in the to_edge_lower_and_transform flow.

Perhaps you could take a look at the similar change for XNNPACK: #13209<https://github.com/pytorch/executorch/pull/13209>

We should also add some unit tests showing: 1) the warning is logged with to_edge, but 2) not logged with to_edge_lower_and_transform.

—
Reply to this email directly, view it on GitHub<https://github.com/pytorch/executorch/pull/15963#issuecomment-3572388215>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BXDRGNQCC4SIIILM7WQWYDL36NMADAVCNFSM6AAAAACNB2NXYGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTKNZSGM4DQMRRGU>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

jagadeeshkaruturi11 committed:
Add call stack inspection to show warning only for deprecated to_edge workflow

Following the pattern from PR #13209 (XNNPACK), this commit:

1. Adds `import inspect` to enable call stack inspection
2. Adds `_check_if_called_from_to_backend()` helper method that:
   - Returns False if called from to_edge_transform_and_lower (recommended flow)
   - Returns True if called from to_backend in deprecated flow
3. Wraps the deprecation warning in a conditional check

This ensures the warning only appears when using the deprecated to_edge() + to_backend() workflow, not when using the recommended to_edge_transform_and_lower() flow.

Unit tests will be added in a separate commit.

jagadeeshkaruturi11 committed:
Add unit tests for deprecation warning

Adds two unit tests to verify the deprecation warning behavior:

1. `test_deprecation_warning_for_to_backend_workflow`: Verifies that the warning IS logged when using the deprecated to_edge() + to_backend() workflow

2. `test_no_warning_for_to_edge_transform_and_lower_workflow`: Verifies that the warning is NOT logged when using the recommended to_edge_transform_and_lower() workflow

Also adds necessary imports:
- import io
- import logging  
- from executorch.exir import to_edge, to_edge_transform_and_lower

These tests ensure the deprecation warning only appears in the deprecated flow, following the same pattern as XNNPACK (PR #13209).

metascroy committed:
Merge branch 'main' into Lokesh9106-patch-1

metascroy commented:
Looks great!  Make you you run the linter though.

To run the linter, execute the following from the executorch repo folder:

```
lintrunner -a
```

This requires you install the lintrunner (pip install -r requirements-lintrunner.txt).

metascroy committed:
Merge branch 'main' into Lokesh9106-patch-1

metascroy committed:
Merge branch 'main' into Lokesh9106-patch-1
